### PR TITLE
Set mac binding age threshold in gateway routers

### PR DESCRIPTION
**Pull request by jcaamano**
On scenarios where the cluster is shut down and started up after some time later, the dynamic mac binding entries in the SBDB might be stale. If any endpoint changed MAC address and sent a GARP, OVN has no chance of processing it if it is shut down.

Set mac_binding_age_threshold to 300 seconds. Most common values used in other implementations range from 60s to 1200s being 300s the most common after a google search.

Unfortunately the mac binding age of an entry is not refreshed on any network event, so entries will be removed from the mac binding table after 300 seconds unconditionaly. There is an initiative to improve this behavior in OVN [1]

This requires also to set mac bindings int the Static_Mac_Binding table instead of the Mac_Binding table so that they are not wiped by the aging mechanism. We also need to clean them up appropriately, otherwise northd may fail to reconcile if they point to a LRP that no longer exists.

[1] https://mail.openvswitch.org/pipermail/ovs-discuss/2023-May/052475.html

### CODE REVIEW

**Comment by jcaamano** [comment]
@dcbw @trozet PTAL

**Comment by jcaamano** [comment]
Saving this flake for later
```
[Fail] Multi node zones interconnect [It] Pod interconnectivity 
/home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/multi_node_zones_interconnect.go:83

[Fail] Multi node zones interconnect [It] Pod interconnectivity 
/home/runner/go/pkg/mod/k8s.io/kubernetes@v1.24.0/test/e2e/framework/pods.go:107
```
https://github.com/ovn-org/ovn-kubernetes/actions/runs/5834359113/job/15825042191?pr=3678

**Comment by jcaamano** [comment]
> Saving this flake for later
> 
> ```
> [Fail] Multi node zones interconnect [It] Pod interconnectivity 
> /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/multi_node_zones_interconnect.go:83
> 
> [Fail] Multi node zones interconnect [It] Pod interconnectivity 
> /home/runner/go/pkg/mod/k8s.io/kubernetes@v1.24.0/test/e2e/framework/pods.go:107
> ```
> 
> https://github.com/ovn-org/ovn-kubernetes/actions/runs/5834359113/job/15825042191?pr=3678

So not a flake after all, consistently failing.

The test that fails checks that two pods on different zones can talk to each other, then brings one pod to the same zone as the other and does the same check and then takes the former pod to its original zone and checks again. The last check fails.

It looks like the pod's node never becomes ready after that last change. ovnkube-node fails finding the output flow for the node's management port. Checking the NBDB, that port is present but not up. It's row is missing from the SBDB Port_Binding table. The port is there in OVS, with ovn-installer=true.

Compare ovn-controller logs before the change
```
2023-08-14T12:55:27.831Z|00018|main|INFO|OVS feature set changed, force recompute.
2023-08-14T12:55:28.960Z|00019|reconnect|INFO|tcp:172.18.0.2:6642: connecting...
2023-08-14T12:55:28.968Z|00020|main|INFO|OVNSB commit failed, force recompute next time.
2023-08-14T12:55:28.968Z|00021|reconnect|INFO|tcp:172.18.0.2:6642: connected
2023-08-14T12:55:28.988Z|00022|binding|INFO|Claiming lport etor-GR_ovn-worker3 for this chassis.
2023-08-14T12:55:28.988Z|00023|binding|INFO|etor-GR_ovn-worker3: Claiming 02:42:ac:12:00:03
2023-08-14T12:55:28.988Z|00024|binding|INFO|Claiming lport rtoe-GR_ovn-worker3 for this chassis.
2023-08-14T12:55:28.988Z|00025|binding|INFO|rtoe-GR_ovn-worker3: Claiming 02:42:ac:12:00:03 172.18.0.3/16
2023-08-14T12:55:37.838Z|00026|memory|INFO|14192 kB peak resident set size after 10.0 seconds
2023-08-14T12:55:37.838Z|00027|memory|INFO|idl-cells-OVN_Southbound:6206 idl-cells-Open_vSwitch:635 lflow-cache-entries-cache-expr:237 lflow-cache-entries-cache-matches:438 lflow-cache-size-KB:889 local_datapath_usage-KB:1 ofctrl_desired_flow_usage-KB:68 ofctrl_installed_flow_usage-KB:50 ofctrl_sb_flow_ref_usage-KB:33
2023-08-14T12:55:58.993Z|00028|memory_trim|INFO|Detected inactivity (last active 30000 ms ago): trimming memory
```
to after
```
2023-08-14T14:18:15.620Z|00018|main|INFO|OVS feature set changed, force recompute.
2023-08-14T14:18:16.784Z|00019|reconnect|INFO|tcp:172.18.0.4:6642: connecting...
2023-08-14T14:18:16.793Z|00020|main|INFO|OVNSB commit failed, force recompute next time.
2023-08-14T14:18:16.793Z|00021|reconnect|INFO|tcp:172.18.0.4:6642: connected
2023-08-14T14:18:17.454Z|00022|binding|INFO|Claiming lport cr-rtos-ovn-worker3 for this chassis.
2023-08-14T14:18:17.454Z|00023|binding|INFO|cr-rtos-ovn-worker3: Claiming 0a:58:0a:f4:02:01 10.244.2.1/24
2023-08-14T14:18:17.514Z|00024|binding|INFO|Claiming lport k8s-ovn-worker3 for this chassis.
2023-08-14T14:18:17.514Z|00025|binding|INFO|k8s-ovn-worker3: Claiming 72:ec:6a:44:e0:e6 10.244.2.2
2023-08-14T14:18:17.515Z|00026|binding|INFO|Claiming lport jtor-GR_ovn-worker3 for this chassis.
2023-08-14T14:18:17.515Z|00027|binding|INFO|jtor-GR_ovn-worker3: Claiming router
2023-08-14T14:18:17.515Z|00028|binding|INFO|Claiming lport rtoe-GR_ovn-worker3 for this chassis.
2023-08-14T14:18:17.515Z|00029|binding|INFO|rtoe-GR_ovn-worker3: Claiming 02:42:ac:12:00:05 172.18.0.5/16
2023-08-14T14:18:17.515Z|00030|binding|INFO|Claiming lport rtoj-GR_ovn-worker3 for this chassis.
2023-08-14T14:18:17.515Z|00031|binding|INFO|rtoj-GR_ovn-worker3: Claiming 0a:58:64:40:00:04 100.64.0.4/16
2023-08-14T14:18:17.515Z|00032|binding|INFO|Claiming lport etor-GR_ovn-worker3 for this chassis.
2023-08-14T14:18:17.515Z|00033|binding|INFO|etor-GR_ovn-worker3: Claiming 02:42:ac:12:00:05
2023-08-14T14:18:17.540Z|00034|binding|INFO|Removing lport k8s-ovn-worker3 ovn-installed in OVS
2023-08-14T14:18:17.700Z|00035|binding|INFO|Claiming lport multi-node-zones-5452_client-pod for this chassis.
2023-08-14T14:18:17.701Z|00036|binding|INFO|multi-node-zones-5452_client-pod: Claiming 0a:58:0a:f4:02:03 10.244.2.3
2023-08-14T14:18:17.707Z|00037|binding|INFO|Removing lport multi-node-zones-5452_client-pod ovn-installed in OVS
2023-08-14T14:18:17.709Z|00038|binding|INFO|Setting lport k8s-ovn-worker3 up in Southbound
2023-08-14T14:18:17.710Z|00039|binding|INFO|Setting lport k8s-ovn-worker3 ovn-installed in OVS
2023-08-14T14:18:17.710Z|00040|binding|INFO|Setting lport multi-node-zones-5452_client-pod ovn-installed in OVS
2023-08-14T14:18:17.710Z|00041|binding|INFO|Setting lport multi-node-zones-5452_client-pod up in Southbound
2023-08-14T14:18:25.632Z|00042|memory|INFO|14736 kB peak resident set size after 10.0 seconds
2023-08-14T14:18:25.632Z|00043|memory|INFO|idl-cells-OVN_Southbound:7207 idl-cells-Open_vSwitch:635 if_status_mgr_ifaces_state_usage-KB:1 if_status_mgr_ifaces_usage-KB:1 lflow-cache-entries-cache-expr:383 lflow-cache-entries-cache-matches:577 lflow-cache-size-KB:1452 local_datapath_usage-KB:1 ofctrl_desired_flow_usage-KB:291 ofctrl_installed_flow_usage-KB:216 ofctrl_sb_flow_ref_usage-KB:116
2023-08-14T14:18:47.708Z|00044|memory_trim|INFO|Detected inactivity (last active 30002 ms ago): trimming memory
```

One of the things I am thinking about is if the way we were previously updating the mac bindings in the SBDB directly (instead of NBDB that this PR does) forces a different kind of recompute in ovn-copntroller.

**Comment by numansiddique** [comment]
@almusil FYI

**Comment by jcaamano** [comment]
@numansiddique helped me here and advised me to delete the static mac bindings on gateway cleanup, something that we were not doing before. On one side, OVN-K should be doing this. On the other side it looks like OVN should be handling this better so perhaps file a bug against OVN as well.

**Comment by jcaamano** [comment]
@JacobTanenbaum I added some logic here to clean the static mac bindings that we do for HO on the NBDB cleanup that was already in place. However, this cleanup only seems to happen on node deletion but not on a node update when the node is missing the appropriate labels or when HO has been disabled. PTAL.

**Comment by jcaamano** [comment]
testing downstream 
https://github.com/openshift/ovn-kubernetes/pull/1823

**Review by tssurya — COMMENTED** [review]
overall i think this change makes sense - it would make more sense if we can wait for ales's fixes to land so that we don't start to suddenly remove bindings after 5mins even if they are being used? I think 23.09 is only a few weeks away right?

**Comment by jcaamano** [comment]
> overall i think this change makes sense - it would make more sense if we can wait for ales's fixes to land so that we don't start to suddenly remove bindings after 5mins even if they are being used? I think 23.09 is only a few weeks away right?

You have more information than I do. How do you know this is already available in 23.09? In regards to waiting, we already discussed renewing the entries every 5 mins shouldn't be a major problem and wouldn't be a functional difference. If there is an unexpected problem, this can give us an early signal. But I don't have a problem with waiting if we think that is better.

File: go-controller/pkg/libovsdb/libovsdb.go
[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/libovsdb/libovsdb.go:1]: > so from this commit onwards we move all bindings to NBDB's static binding table and stop directly creating stuff in SBDB? Is there a migration path for existing bindings to be wiped off during update from the old table? to me it seems like old bindings will continue to be there in sbdb's mac binding table and then new ones will start to be created in nbdb? Shouldn't we take care of upgrades?

The aging mechanism will get rid of the old mac bindings. Before that, the new mac bindings will overwrite the old mac bindings which are, in any case, the same bindings anyway.
 
> Also does this mean moving forward OVN handles the mac binding logic totally? -> So as of today, we go with the timestamps and forcefully prune all bindings older than 300s whether they are active or not, and once ales's fixes land, in 23.09, we stop flushing stupidly and instead use the idle age logic?

yes, but I don't know if that's available in 23.09 already.

File: go-controller/pkg/ovn/gateway/gateway.go [hunk 1]
```diff
@@ -87,3 +87,21 @@ func CreateDummyGWMacBindings(nbClient libovsdbclient.Client, nodeName string) e
 
 	return nil
 }
+
+// DeleteDummyGWMacBindings removes mac bindings (ipv4 and ipv6) for a fake next hops
+// used by host->service traffic
+func DeleteDummyGWMacBindings(nbClient libovsdbclient.Client, nodeName string) error {
```
[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/ovn/gateway/gateway.go:93]: northd is not cleaning up the mac binding for us when the logical port is gone, this might be a northd bug though

File: go-controller/pkg/ovn/hybrid.go [hunk 2]
```diff
@@ -307,6 +308,18 @@ func (oc *DefaultNetworkController) removeHybridLRPolicySharedGW(nodeName string
 	}); err != nil {
 		return fmt.Errorf("failed to delete static route %s from %s, error: %v", name+"gr", ovntypes.GWRouterPrefix+nodeName, err)
 	}
+
+	if node.Annotations[hotypes.HybridOverlayDRIP] != "" {
```
[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/ovn/hybrid.go:312]: what specifically? we were not removing the mac bindings before, but we should have I think

File: go-controller/pkg/ovn/master_test.go
[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/ovn/master_test.go:1]: can you point me to the specific tests?

**Comment by jcaamano** [comment]
> > overall i think this change makes sense - it would make more sense if we can wait for ales's fixes to land so that we don't start to suddenly remove bindings after 5mins even if they are being used? I think 23.09 is only a few weeks away right?
> 
> You have more information than I do. How do you know this is already available in 23.09? In regards to waiting, we already discussed renewing the entries every 5 mins shouldn't be a major problem and wouldn't be a functional difference. If there is an unexpected problem, this can give us an early signal. But I don't have a problem with waiting if we think that is better.

I see that the mailing list mentions 23.09 but I read that as a plan not as a fact.

File: go-controller/pkg/ovn/master_test.go
[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/ovn/master_test.go:1]: Note that those tsts generally re-use `generateGatewayInitExpectedNB` which includes the new mac bindings

[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/ovn/master_test.go:1]: ah ack ok hence the confusion (on the commit itself I saw you deleting the sbdb stuff but didn't see us ensuring the expected data for the new stuff). consider this comment resolved.

File: go-controller/pkg/ovn/hybrid.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/ovn/hybrid.go:312]: i meant mac bindings -> ack.

**Comment by tssurya** [comment]
> 
> You have more information than I do. How do you know this is already available in 23.09?

@almusil said its in 23.09 


> In regards to waiting, we already discussed renewing the entries every 5 mins shouldn't be a major problem and wouldn't be a functional difference. If there is an unexpected problem, this can give us an early signal. But I don't have a problem with waiting if we think that is better.

I think the early signal makes sense :) won't stop this PR from merging.

File: go-controller/pkg/libovsdb/libovsdb.go
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/libovsdb/libovsdb.go:1]: > > so from this commit onwards we move all bindings to NBDB's static binding table and stop directly creating stuff in SBDB? Is there a migration path for existing bindings to be wiped off during update from the old table? to me it seems like old bindings will continue to be there in sbdb's mac binding table and then new ones will start to be created in nbdb? Shouldn't we take care of upgrades?
> 
> The aging mechanism will get rid of the old mac bindings. Before that, the new mac bindings will overwrite the old mac bindings which are, in any case, the same bindings anyway.
> 

ah I see so:
when you create the new stuff in nbdb northd will update the stuff a.k.a overwrite it in SBDB.
and after 5mins they will get flushed anyways and that means SBDB will also have the same flush. ok gotcha makes sense. (resolve)

**Review by tssurya — APPROVED** [review]
/lgtm

File: go-controller/pkg/libovsdb/util/mac_binding.go [hunk 3]
```diff
@@ -48,3 +50,37 @@ func CreateMACBinding(sbClient libovsdbclient.Client, logicalPort, datapathName
...
+				Function: ovsdb.ConditionEqual,
+				Value:    ips[i],
+			},
+		)
+	}
+
+	// Delete using the client native API instead of model client to avoid
+	// having to monitor & cache the SBDB mac binding table.
+	// This will encode the conditions as provided and send them to the server
+	// if there is no cache. The operation is idempotent and there is no error
```
[the remark this answers is not in the document]
Comment (trozet) [reply go-controller/pkg/libovsdb/util/mac_binding.go:32]: would have been helpful to include this explanation in the commit message for people that dont know what a native api is

[the remark this answers is not in the document]
Comment (trozet) [reply go-controller/pkg/libovsdb/util/mac_binding.go:32]: wait a second...dont we block the creation of the sbClient on the monitor itself. In other words, we cant have the client unless monitor worked correctly and has already populated the cache. I thought we do this intentionally to avoid a situation where we execute client commands with a cache that is not populated. If that is true, then how does this native strategy make any difference? I would think it would only matter if a client was created without monitoring.

File: go-controller/pkg/ovn/gateway_cleanup.go [hunk 4]
```diff
@@ -256,10 +256,7 @@ func (oc *DefaultNetworkController) cleanupDGP(nodes *kapi.NodeList) error {
 		oc.delPbrAndNatRules(node.Name, []string{types.InterNodePolicyPriority, types.MGMTPortPolicyPriority})
 	}
 	// remove SBDB MAC bindings for DGP
-	p := func(item *sbdb.MACBinding) bool {
-		return item.IP == types.V4NodeLocalNATSubnetNextHop || item.IP == types.V6NodeLocalNATSubnetNextHop
-	}
-	err := libovsdbops.DeleteMacBindingWithPredicate(oc.sbClient, p)
+	err := libovsdbutil.DeleteSbdbMacBindingsWithIPs(oc.sbClient, types.V4NodeLocalNATSubnetNextHop, types.V6NodeLocalNATSubnetNextHop)
```
[the remark this answers is not in the document]
Comment (trozet) [reply go-controller/pkg/ovn/gateway_cleanup.go:265]: yes we should just remove this code now.

File: go-controller/pkg/libovsdb/libovsdb.go
[the remark this answers is not in the document]
Comment (trozet) [reply go-controller/pkg/libovsdb/libovsdb.go:1]: I think if you had written this information in the commit message it would make it more obvious.

File: go-controller/pkg/ovn/gateway/gateway.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (trozet) [reply go-controller/pkg/ovn/gateway/gateway.go:93]: @dceara @numansiddique is this a bug?

File: go-controller/pkg/types/const.go [hunk 5]
```diff
@@ -202,4 +202,9 @@ const (
 	// InformerSyncTimeout is used to wait from the initial informer cache sync.
 	// It allows ~4 list() retries with the default reflector exponential backoff config
 	InformerSyncTimeout = 20 * time.Second
+
+	// GRMacBindingAgeThreshold is the lifetime in seconds of each MAC binding
+	// entry for the gateway routers. After this time, the entry is removed and
+	// may be refreshed with a new ARP request.
+	GRMacBindingAgeThreshold = "300"
```
Comment (trozet) [inline go-controller/pkg/types/const.go:209]: this should be MAC

**Review by trozet — CHANGES_REQUESTED** [review]
I'd like to see the commit messages updated with proper commit body and explanation of what the commit is doing.

File: go-controller/pkg/ovn/gateway/gateway.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dceara) [reply go-controller/pkg/ovn/gateway/gateway.go:93]: The schema for NB. (static) mac bindings is very relaxed:
```
        "Static_MAC_Binding": {
            "columns": {
                "logical_port": {"type": "string"},
                "ip": {"type": "string"},
                "mac": {"type": "string"},
                "override_dynamic_mac": {"type": "boolean"}},
            "indexes": [["logical_port", "ip"]],
             "isRoot": true},
```

The schema for the SB mac binding table is very similar.  The only difference is that ovn-northd doesn't expect anyone else to write to the SB directly (ovn-kube did but that's not really the right way to interact with OVN) so ovn-northd is in charge of creating/deleting SB.MAC_Bindings.

For the NB, it's a bit different, ovn-northd didn't create the NB.Static_MAC_Binding record, it's not clear whether it should clean it or not.

There's always the possibility that the CMS/user/admin/etc creates the Static_MAC_Binding before creating the LRP.  The schema allows that.  If northd were to automatically delete all "stale" Static_MAC_Bindings, this use case would break.

TL/DR: I don't really think this is a bug.  NB records are usually not owned by ovn-northd.

File: go-controller/pkg/libovsdb/util/mac_binding.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/libovsdb/util/mac_binding.go:32]: Not sure I understand the concern. Perhaps I have not been clear enough with the comment.

In this PR, we no longer monitor the SBDB MAC Binding table, so we won't have it in the cache and thus we can use model client API as that relies on the cache: it looks up the model and if it doesn't find any, it won't send the delete towards the server.

As opposed to the native client API, which will encode and send the delete ops to the server as they have been provided to it regardless of whether the SBDB MAC Binding table is being monitored or not.

File: go-controller/pkg/libovsdb/libovsdb.go
[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/libovsdb/libovsdb.go:1]: Will add it to the commit message.

File: go-controller/pkg/ovn/gateway_cleanup.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/ovn/gateway_cleanup.go:265]: should we do this in this PR?

[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/ovn/gateway_cleanup.go:265]: Also note that the cleanup is version dependent (not done unconditionally)
```
	// If version is less than Host -> Service with OpenFlow, we need to remove and cleanup DGP
	if err == nil && ((ver < types.OvnHostToSvcOFTopoVersion && config.Gateway.Mode == config.GatewayModeShared) ||
		(ver < types.OvnRoutingViaHostTopoVersion)) {
		err = oc.cleanupDGP(existingNodes)
	}
```

File: go-controller/pkg/ovn/gateway/gateway.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/ovn/gateway/gateway.go:93]: @dceara 

While the NB records are usually not owned by ovn-northd, we would have a different behavior with the referential integrity that could have happened if `logical_port` was a strong reference and `Static_MAC_Binding` was a non root table.

I am not sure if there is a real use case behind the current schema design or if this has been perhaps an oversight when copying the SBDB table definition to NBDB which might have more implications.

One of those implications, and probably the real bug, is that northd fails to reconcile if `logical_port` does not exist which probably should be handled more gracefully with the current schema design.

[the remark this answers is not in the document]
Comment (dceara) [reply go-controller/pkg/ovn/gateway/gateway.go:93]: It's not OK to add the strong reference now.  We'd potentially break users.  I suspect there are users.  This feature was developed upstream and on a community member request: https://github.com/ovn-org/ovn/commit/b22684d4e37c3dd205a959925b6e3f1827bbe045


> One of those implications, and probably the real bug, is that northd fails to reconcile if `logical_port` does not exist which probably should be handled more gracefully with the current schema design.

We cannot assume that the NB entry should be removed.  With the current schema the user is allowed to insert NB.Static_MAC_Binding and NB.Logical_Router_Port in any order they wish.

File: go-controller/pkg/libovsdb/util/mac_binding.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (trozet) [reply go-controller/pkg/libovsdb/util/mac_binding.go]: yeah it made more sense as I got to the other commits. I think the initial commit message is what was confusing.

File: go-controller/pkg/ovn/gateway_cleanup.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (trozet) [reply go-controller/pkg/ovn/gateway_cleanup.go:265]: I think this was due to upgrade constraints at the time.

**Review by trozet — APPROVED** [review]
